lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -391,20 +478,30 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
       markOverdefined(verbatim.getResult());
     else if (auto verbatim = dyn_cast<VerbatimWireOp>(op))
       markOverdefined(verbatim.getResult());
+
+    // This tracks a dependency from field refs to operations which need to be
+    // added to worklist when lattice values change.
+    if (!isAggregate(&op)) {
+      for (auto operand : op.getOperands()) {
+        auto fieldRef = getOrCacheFieldRefFromValue(operand);
+        auto firrtlType = operand.getType().dyn_cast<FIRRTLType>();
+        if (!firrtlType)
+          continue;
+        foreachFIRRTLGroundType(firrtlType, [&](unsigned fieldID, auto type) {
+          fieldRefToUsers[fieldRef.getSubField(fieldID)].push_back(&op);
```

Comment:
Does this result in same operation being added multiple times for a fieldref if the operation uses the same operand repeatedly?  Consider whether avoiding this or unique'ing  at a later point would be better than any resulting costs re:mem/redundant processing.

Reply:
> Does this result in same operation being added multiple times for a fieldref if the operation uses the same operand repeatedly?

Good point, yes. Currently `TinyPtrVector` is used but I'll try MapVector to ensure the uniqueness. Note that `value.getUsers()` could also return same operations so the current implementation should match the previous behavior but I think this looks good improvement, thanks! :+1:

Reply:
I tried SmallMapVector but it was 5~10% slower than TinyPtrVector (possibly because practically there is not many duplication and TinyPtrVector is more efficient when iterating small vectors) so I'll continue with the current implementation. Thank you for the suggestion!